File: src/day21.rs
#[allow(dead_code)]
pub fn solution() {
    println!("Part 1: {}", deterministic_score(8, 7));
}

/// Plays a game of dirac dice with a deterministic d100 (rolls 1, then 2, ...), and returns
/// the score of the losing player by the number of times the die was rolled during the game.
/// Dirac dice is played by rolling a die 3 times, then moving a pawn that many times
/// around a board marked 1..=10.  The square the player lands on is added to the player's score.
/// The game ends as a win for any player whose score reaches at least 1000.
fn deterministic_score(player1_start: i64, player2_start: i64) -> i64 {
    let mut player1_score = 0;
    let mut player1_space = player1_start;

    let mut player2_score = 0;
    let mut player2_space = player2_start;

    let mut die = DeterministicD100::default();
    let mut rolls = 0;

    loop {
        rolls += 3;
        if turn(&mut die, &mut player1_space, &mut player1_score) {
            return rolls * player2_score;
        }

        rolls += 3;
        if turn(&mut die, &mut player2_space, &mut player2_score) {
            return rolls * player1_score;
        }
    }
}

fn turn(die: &mut DeterministicD100, space: &mut i64, score: &mut i64) -> bool {
    *space = (((*space - 1) + die.next() + die.next() + die.next()) % 10) + 1;
    *score += *space;

    *score >= 1000
}

#[derive(Debug)]
struct DeterministicD100 {
    next: i64,
}

impl Default for DeterministicD100 {
    fn default() -> Self {
        DeterministicD100 { next: 1 }
    }
}

impl DeterministicD100 {
    fn next(&mut self) -> i64 {
        let value = self.next;
        self.next = self.next % 100 + 1;
        value
    }
}

#[test]
fn deterministic_sample() {
    assert_eq!(739785, deterministic_score(4, 8));
}

File: internal/day21/day21.go
package day21

// Part1 plays a game of dirac dice with a deterministic d100 (rolls 1, then 2, ...), and returns
// the score of the losing player by the number of times the die was rolled during the game.
// Dirac dice is played by rolling a die 3 times, then moving a pawn that many times
// around a board marked [1,10].  The square the player lands on is added to the player's score.
// The game ends as a win for any player whose score reaches at least 1000.
func Part1(player1Start, player2Start int) int {
	players := []player{
		{space: player1Start},
		{space: player2Start},
	}

	turn := 0
	die := d100{next: 1}

	for players[0].score < 1000 && players[1].score < 1000 {
		players[turn].move(die.roll(3))
		turn = (turn + 1) % len(players)
	}

	return min(players[0].score, players[1].score) * die.rolls
}

type player struct {
	space int
	score int
}

// move moves this player's pawn around a board marked [1,10] and adds the square the player lands on to it's score.
func (p *player) move(roll int) {
	p.space = (p.space-1+roll)%10 + 1
	p.score += p.space
}

// d100 is a deterministic d100 that rolls 1, then 2, ..., then 100, then 1, etc.
type d100 struct {
	next  int
	rolls int
}

// roll rolls the die the given number of times, returning the sum of the rolls.
func (die *d100) roll(times int) int {
	die.rolls += times

	value := 0

	for i := 0; i < times; i++ {
		value += die.next
		die.next = die.next%100 + 1
	}

	return value
}
